### PR TITLE
Database performance improvements

### PR DESCRIPTION
**Pull request by jfberry**
image

 * Delay writing wild pokemon entries from GMO, with the expectation an encounter will overtake it; if it doesn't arrive within 15 seconds then write
 * Optimise reads for new pokemon records so avoid the check being handled more than once

New config options allow turning off wild pokemon, and nearby pokemon processing (see README)
Advice on optimising mariadb

### CODE REVIEW

File: decoder/pokemon.go [hunk 1]
```diff
@@ -489,6 +510,15 @@ func (pokemon *Pokemon) calculateIv(a int64, d int64, s int64) {
 	pokemon.Iv = null.FloatFrom(float64(a+d+s) / .45)
 }
 
+// wildSignificantUpdate returns true if the wild pokemon is significantly different from the current pokemon and
```
Comment (Mygod) [inline decoder/pokemon.go:513]: Forgot to change this? Also this func seems unused.

File: decoder/main.go [hunk 2]
```diff
@@ -252,35 +252,62 @@ func UpdateFortBatch(ctx context.Context, db db.DbDetails, p []RawFortData) {
...
-			log.Printf("getOrCreatePokemonRecord: %s", err)
-		} else {
-			pokemon.updateFromWild(ctx, db, wild.Data, int64(wild.Cell), int64(wild.Timestamp), username)
-			savePokemonRecord(ctx, db, pokemon)
-		}
+			pokemon, err := getOrCreatePokemonRecord(ctx, db, encounterId)
+			if err != nil {
+				log.Errorf("getOrCreatePokemonRecord: %s", err)
+			} else {
+				if pokemon == nil || pokemon.isNewRecord() || pokemon.wildSignificantUpdate(wild.Data) {
```
Comment (Mygod) [inline decoder/main.go:265]: No need to check for `nil` here. :)

File: decoder/pokemon.go [hunk 3]
```diff
@@ -417,6 +429,15 @@ func (pokemon *Pokemon) addWildPokemon(ctx context.Context, db db.DbDetails, wil
 	pokemon.SpawnId = null.IntFrom(spawnId)
 }
 
+// wildSignificantUpdate returns true if the wild pokemon is significantly different from the current pokemon and
+// should be written.
+func (pokemon *Pokemon) wildSignificantUpdate(wildPokemon *pogo.WildPokemonProto) bool {
+	return pokemon.SeenType.ValueOrZero() == SeenType_Cell ||
+		pokemon.PokemonId != int16(wildPokemon.Pokemon.PokemonId) ||
+		pokemon.Form.ValueOrZero() != int64(wildPokemon.Pokemon.PokemonDisplay.Form) ||
+		pokemon.Weather.ValueOrZero() != int64(wildPokemon.Pokemon.PokemonDisplay.WeatherBoostedCondition)
```
Comment (Mygod) [inline decoder/pokemon.go:438]: How about `Costume` and `Gender`? `SeenType_Nearby`? Spawnpoint update?

Comment (Mygod) [reply decoder/pokemon.go:438]: Also expire timer is more accurate in wild. Hmm..

File: decoder/main.go [hunk 4]
```diff
@@ -252,35 +252,62 @@ func UpdateFortBatch(ctx context.Context, db db.DbDetails, p []RawFortData) {
...
+					updateTime := time.Now().Unix()
```
Comment (Mygod) [inline decoder/main.go:266]: Also can use `timestampMs` here.

**Comment by ccev** [comment]
@jfberry i think you forgot to update the example config?

**Comment by jfberry** [comment]
> @jfberry i think you forgot to update the example config?

it is in the README documentation, I didn't want it in the example config as the defaults as fine for most people

**Pull request merged**